pgx/src/datum/range.rs

```diff
@@ -9,56 +9,281 @@ Use of this source code is governed by the MIT license that can be found in the
 
 //! Utility functions for working with `pg_sys::RangeType` structs
 use crate::{
-    pg_sys, void_mut_ptr, AnyNumeric, Date, FromDatum, IntoDatum, Numeric, Timestamp,
-    TimestampWithTimeZone,
+    pg_sys, AnyNumeric, Date, FromDatum, IntoDatum, Numeric, Timestamp, TimestampWithTimeZone,
 };
-use pgx_pg_sys::{Oid, RangeBound};
 use pgx_sql_entity_graph::metadata::{
     ArgumentError, Returns, ReturnsError, SqlMapping, SqlTranslatable,
 };
-use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut, RangeFrom, RangeInclusive, RangeTo, RangeToInclusive};
+
+/// A Postgres Range's "lower" or "upper" value
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub enum RangeBound<T> {
+    Infinite,
+    Inclusive(T),
+    Exclusive(T),
+}
 
-/// Represents Datum to serialized RangeType PG struct
-pub struct Range<T: FromDatum + IntoDatum + RangeSubType> {
-    ptr: *mut pg_sys::varlena,
-    range_type: *mut pg_sys::RangeType,
-    _marker: PhantomData<T>,
+impl<T> RangeBound<T>
+where
+    T: RangeSubType,
+{
+    /// Returns a reference to this [`RangeBound`]'s underlying value.
+    ///
+    /// If this [`RangeBound`] is [`RangeBound::Infinite`], then `None` is returned, otherwise
+    /// `Some(&T)`
+    #[inline]
+    pub fn get(&self) -> Option<&T> {
+        match self {
+            RangeBound::Infinite => None,
+            RangeBound::Inclusive(v) | RangeBound::Exclusive(v) => Some(v),
+        }
+    }
+
+    /// Convert this pgx [`RangeBound`] into the equivalent Postgres [`pg_sys::RangeBound`].
+    ///
+    /// Note that the `lower` property is always set to false as a [`RangeBound`] doesn't know the
+    /// end on which it's placed.
```

Comment:
The other question is if we should be exposing an `into_pg` that allows constructing a `RangeBound` that doesn't actually know whether it is the `lower` range or not yet says it is not, instead of requiring people to use a similar function on the larger `Range`. I don't know if it's common to just manipulate `RangeBound`s separately through Postgres functions, though.

Reply:
we can make that function private.  It's used in the `IntoDatum` impl.